### ISSUE TITLE
The anchor point of markdown file is incorrect after using vue component

### ISSUE THREAD

**Issue by newives** [vuejs/vitepress#3428]
### Describe the bug

The anchor point of markdown file is incorrect after using vue component

### Reproduction

https://stackblitz.com/edit/vite-ntyrly?file=docs%2Fexample.md

### Expected behavior

Anchor point correct

### System Info

```Text
mac
Firefox
```

### Additional context

<img width="1409" alt="图片" src="https://github.com/vuejs/vitepress/assets/81661030/3c1e94e7-fd9e-4d2b-8809-c2c6396c97a4">

### Validations

- [X] Check if you're on the [latest VitePress version](https://github.com/vuejs/vitepress/releases/latest).
- [X] Follow our [Code of Conduct](https://vuejs.org/about/coc.html)
- [X] Read the [docs](https://vitepress.dev).
- [X] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.

**Comment by Octobug** [comment]
I've checked your live demo.

It doesn't seem like a bug caused by using a vue component. The issue appears when there isn't enough height for the page. Even zooming for the browser can cause it.

If you insert enough contents in the `<test />` vue components, it works fine.

<img width="1377" alt="image" src="https://github.com/vuejs/vitepress/assets/8007022/84b3fe50-4de7-41cb-9bb2-3e94e3af930d">

For example, in this vitepress doc page, you zoom out `-` the browser, the same behavior occurs: https://vitepress.dev/reference/default-theme-last-updated#site-level-config

**Comment by newives** [comment]
<img width="1365" alt="图片" src="https://github.com/vuejs/vitepress/assets/81661030/27f6441d-a582-4b10-bb76-90aba82b5534">
I'm sure this is a bug
My updated demo

https://stackblitz.com/edit/vite-ntyrly?file=docs%2Ftest.vue

**Comment by Octobug** [comment]
> <img alt="图片" width="1365" src="https://private-user-images.githubusercontent.com/81661030/295089366-27f6441d-a582-4b10-bb76-90aba82b5534.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ3NzIyOTEsIm5iZiI6MTcwNDc3MTk5MSwicGF0aCI6Ii84MTY2MTAzMC8yOTUwODkzNjYtMjdmNjQ0MWQtYTU4Mi00YjEwLWJiNzYtOTBhYmE4MmI1NTM0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTA5VDAzNDYzMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI1MmMwNzEyNTI2Y2I2ZWZiMGZiNDVlOTdjOThlMjQxN2ZmMjk2ZjAyOTFmYzU2OWQ2YzA1NGZhYmIyMzRhNjImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.tfhUjgc8zozjr014H17z97Nrdz8YiC4ZH6ZE1fiqwaA"> I'm sure this is a bug My updated demo
> https://stackblitz.com/edit/vite-ntyrly?file=docs%2Ftest.vue

How do you get to the `#powered-by-vite` URL? Click on the outline item, click on the `#` in front of the "Powered by Vite", or access directly through a copied URL? @newives

**Comment by newives** [comment]
Click on `Powered by Vite` on the left and refresh the webpage to reproduce

**Comment by brc-dd** [comment]
Ah right, it's not maintaining position on reload.

**Comment by Octobug** [comment]
> Click on `Powered by Vite` on the left and refresh the webpage to reproduce

Strange, I cannot reproduce the bug. Unless I

Click on `Powered by Vite` on the left, ***then scroll to somewhere else***, and refresh the webpage to reproduce.

**Comment by brc-dd** [comment]
> Strange, I cannot reproduce the bug. 

Seems to be only happening on firefox.

**Comment by newives** [comment]
The demo below does not locate the accurate anchor point in any browser.

https://stackblitz.com/edit/vite-fjgh8q?file=docs%2Fexample.md

<img width="1542" alt="图片" src="https://github.com/vuejs/vitepress/assets/81661030/24c5fc54-43ea-4ff6-aa41-29b02018b840">

**Comment by brc-dd** [comment]
> The demo below does not locate the accurate anchor point in any browser.

That's just layout shift. You haven't added height/width or aspect ration to that image - https://web.dev/articles/optimize-cls

**Comment by Octobug** [comment]
> > The demo below does not locate the accurate anchor point in any browser.
> 
> That's just layout shift. You haven't added height/width or aspect ration to that image - https://web.dev/articles/optimize-cls
> 
> Add that and it will work fine in Chromium/Safari.

I've read this part of the article [Modern best practice for setting image dimensions](https://web.dev/articles/optimize-cls#modern_best_practice_for_setting_image_dimensions). It looks a bit inconvenient for writing a markdown post with images...

Could it be done by postponing the execution of `scrollTo()`, something like:

```js
window.onload = () => {
  scrollTo(...);
}
```

But with `window.onload` it might be a problem for `loading="lazy"` images, [When window.onload Triggers With Lazy Loaded Images](https://promincproductions.com/blog/window-onload-with-lazy-loaded-images/)

**Comment by brc-dd** [comment]
> It looks a bit inconvenient for writing a markdown post with images

You can use some markdown-it plugin like https://github.com/boyum/markdown-it-image-size or https://github.com/ulfschneider/markdown-it-fitmedia though. There are also vite plugins out there that can optimize the images (create modern formats or blurhash like fallbacks).

**Comment by zhangyx1998** [comment]
> > > The demo below does not locate the accurate anchor point in any browser.
> > 
> > 
> > That's just layout shift. You haven't added height/width or aspect ration to that image - https://web.dev/articles/optimize-cls
> > Add that and it will work fine in Chromium/Safari.
> 
> I've read this part of the article [Modern best practice for setting image dimensions](https://web.dev/articles/optimize-cls#modern_best_practice_for_setting_image_dimensions). It looks a bit inconvenient for writing a markdown post with images...
> 
> Could it be done by postponing the execution of `scrollTo()`, something like:
> 
> ```js
> window.onload = () => {
>   scrollTo(...);
> }
> ```
> 
> This should also solve the Firefox problem. The incorrect refresh scroll is done by firefox automatically. It seems that Chrome also does it before `scrollTo()`, so in Chrome it works fine.
> 
> But with `window.onload` it might be a problem for `loading="lazy"` images, [When window.onload Triggers With Lazy Loaded Images](https://promincproductions.com/blog/window-onload-with-lazy-loaded-images/)

I think you need to take into account the scenario when user scrolls away before `onload` fires, otherwise the page will jump back even of user has already scrolled away.

That means you also need to listen for `onscroll`, and disable `onload` behavior upon user scroll.

**Comment by Octobug** [comment]
> I think you need to take into account the scenario when user scrolls away before `onload` fires, otherwise the page will jump back even if user has already scrolled away.
> 
> That means you also need to listen for `onscroll`, and disable `onload` behavior upon user scroll.

Yeah, my thought was to postpone the execution of `scrollTo()`. I wasn't sure in what way it could be done perfectly.

**Comment by newives** [comment]
> > The demo below does not locate the accurate anchor point in any browser.
> 
> That's just layout shift. You haven't added height/width or aspect ratio to that image - [web.dev/articles/optimize-cls](https://web.dev/articles/optimize-cls)
> 
> Add that and it will work fine in Chromium/Safari.

In this case, each one needs to be set high. Is it possible to provide an API similar to the following?
Make sure the content is loaded before scrolling to the anchor point
```js
window.onload = function() {
  if (window.location.hash) {
    const hash = window.location.hash;
    const targetElement = document.querySelector(hash);
    if (targetElement) {
      window.scrollTo(0, targetElement.offsetTop);
    }
  }
};
```

**Comment by newives** [comment]
After sorting out the scenarios that will affect the anchor point, it seems that only fixed height is the best way at present.

1. Dynamic Content Loading: When a page contains asynchronously loaded content, such as data or components fetched via AJAX requests, this content can alter the document flow after the initial page load, thus affecting the positioning of anchor points.

2. Image and Other Resource Load Times: Large images and other media resources may require additional time to load, which can cause the page layout to change after the initial positioning.

3. CSS Transitions and Animations: If CSS transitions or animations are used during page load, these effects can temporarily alter the position or dimensions of elements, thereby affecting the accurate positioning of anchors.

4. User Custom Browser Settings: Users may have set zoom levels or other display options in their browsers, which can also affect element positioning.

5. Browser Compatibility and Behavioral Differences: Different browsers may handle anchor navigation differently, especially when dealing with complex layouts and dynamic content.

**Comment by svanimpe** [comment]
I had a similar issue. In my case, it was caused by a Vue component that added an `<img>` but that image didn't have a width and height specified. Adding those resolved the issue.
